src/librbd/operation/DisableFeaturesRequest.cc

```diff
@@ -0,0 +1,655 @@
+
+  if (m_mirror_mode != cls::rbd::MIRROR_MODE_POOL) {
+    send_acquire_exclusive_lock();
+    return nullptr;
+  }
+
+  // mark mirroring as disabling and prune all sync snapshots
+  // before acquiring locks
+  send_disable_mirror_image();
+  return nullptr;
+}
+
+template <typename I>
+void DisableFeaturesRequest<I>::send_disable_mirror_image() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  Context *ctx = create_context_callback<
+    DisableFeaturesRequest<I>,
+    &DisableFeaturesRequest<I>::handle_disable_mirror_image>(this);
+
+  image::MirrorDisableRequest<I> *req =
+    image::MirrorDisableRequest<I>::create(&image_ctx, false, false, ctx);
+  req->send();
+}
+
+template <typename I>
+Context *DisableFeaturesRequest<I>::handle_disable_mirror_image(int *result) {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(cct) << "failed to disable image mirroring: " << cpp_strerror(*result)
+               << dendl;
+    return handle_finish(*result);
+  }
+
+  send_acquire_exclusive_lock();
+  return nullptr;
+}
+
+
+template <typename I>
+void DisableFeaturesRequest<I>::send_acquire_exclusive_lock() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  if ((m_features & RBD_FEATURE_JOURNALING) != 0) {
+    RWLock::WLocker snap_locker(image_ctx.snap_lock);
+    image_ctx.set_journal_policy(new journal::DisabledPolicy());
+    m_disabling_journal = true;
+  }
+
+  Context *ctx = create_context_callback<
+    DisableFeaturesRequest<I>,
+    &DisableFeaturesRequest<I>::handle_acquire_exclusive_lock>(this);
+
+  {
+    RWLock::WLocker locker(image_ctx.owner_lock);
+    // if disabling features w/ exclusive lock supported, we need to
+    // acquire the lock to temporarily block IO against the image
+    if (image_ctx.exclusive_lock != nullptr &&
+	!image_ctx.exclusive_lock->is_lock_owner()) {
+      m_acquired_lock = true;
+
+      image_ctx.exclusive_lock->request_lock(ctx);
+      return;
+    }
+  }
+
+  ctx->complete(0);
+}
+
+template <typename I>
+Context *DisableFeaturesRequest<I>::handle_acquire_exclusive_lock(int *result) {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(cct) << "failed to lock image: " << cpp_strerror(*result) << dendl;
+    return handle_finish(*result);
+  } else if (m_acquired_lock && (image_ctx.exclusive_lock == nullptr ||
+				 !image_ctx.exclusive_lock->is_lock_owner())) {
+    lderr(cct) << "failed to acquire exclusive lock" << dendl;
+    *result = -EROFS;
+    return handle_finish(*result);
+  }
+
+  if ((m_features & ~image_ctx.features) != 0) {
```

Comment:
This check can only be invoked from the synchronous "Operations::update_features" function since we need to preserve idempotency on image watcher RPC or journal event replay. Same comment for enabling features.